File: src/agent_feed/uninstall.py
"""Safe removal of Agent Feed-managed project assets."""

from __future__ import annotations

import shutil
from pathlib import Path

from agent_feed.adapters import claude, cursor
from agent_feed.asset_trust import project_trust_uninstall_plan, remove_project_trust_state
from agent_feed.fs import same_tree
from agent_feed.models import WriteAction


def uninstall_plan(root: Path, *, dry_run: bool) -> list[WriteAction]:
    action = "would delete" if dry_run else "delete"
    actions: list[WriteAction] = []

    _plan_file(
        actions,
        root / "CLAUDE.md",
        action=action,
        safe=claude.is_managed_claude_md(root / "CLAUDE.md"),
        unmanaged_detail="unmanaged Claude instructions; not removed",
        managed_detail="managed Claude adapter",
    )

    claude_root = root / ".claude"
    claude_skills = claude_root / "skills"
    claude_readme = claude_root / "README.md"
    has_managed_claude_skills = claude.is_managed_skill_mirror(claude_root)
    _plan_path(
        actions,
        claude_skills,
        action=action,
        safe=has_managed_claude_skills,
        unmanaged_detail="unmanaged Claude skills; not removed",
        managed_detail="managed Claude skills mirror",
    )
    _plan_file(
        actions,
        claude_readme,
        action=action,
        safe=has_managed_claude_skills,
        unmanaged_detail="unmanaged Claude README; not removed",
        managed_detail="managed Claude skills README",
    )

    cursor_rule = root / ".cursor/rules/agent-feed.mdc"
    _plan_file(
        actions,
        cursor_rule,
        action=action,
        safe=cursor.is_managed_cursor_rule(cursor_rule),
        unmanaged_detail="unmanaged Cursor rule; not removed",
        managed_detail="managed Cursor adapter",
    )

    codex_skills = root / ".codex/skills"
    agents_skills = root / ".agents/skills"
    _plan_path(
        actions,
        codex_skills,
        action=action,
        safe=agents_skills.is_dir()
        and codex_skills.is_dir()
        and same_tree(agents_skills, codex_skills),
        unmanaged_detail="legacy .codex/skills is not a verified mirror; not removed",
        managed_detail="legacy Agent Feed Codex skill mirror",
    )

    agents_md = root / "AGENTS.md"
    _plan_file(
        actions,
        agents_md,
        action=action,
        safe=is_agent_feed_agents_md(agents_md),
        unmanaged_detail="unmanaged AGENTS.md; not removed",
        managed_detail="Agent Feed entry contract",
    )

    agents_dir = root / ".agents"
    _plan_path(
        actions,
        agents_dir,
        action=action,
        safe=is_agent_feed_agents_dir(agents_dir),
        unmanaged_detail="unmanaged .agents directory; not removed",
        managed_detail="Agent Feed protocol directory",
    )

    actions.extend(project_trust_uninstall_plan(root, dry_run=dry_run))
    return actions


def apply_uninstall_plan(root: Path, actions: list[WriteAction]) -> list[WriteAction]:
    applied: list[WriteAction] = []
    for item in actions:
        if item.action != "delete":
            continue
        if not item.path.exists():
            applied.append(WriteAction(item.path, "skip", "already absent"))
            continue
        if item.path.is_dir():
            shutil.rmtree(item.path)
        else:
            item.path.unlink()
        applied.append(WriteAction(item.path, "deleted", item.detail))

    for directory in {
        action.path.parent
        for action in actions
        if action.action == "delete"
        and action.path.name in {"agent-feed.mdc", "README.md", "skills"}
    }:
        remove_empty_parents(directory)

    applied.extend(remove_project_trust_state(root))
    return applied


def has_deletions(actions: list[WriteAction]) -> bool:
    return any(
        action.action in {"delete", "would delete", "update", "would update"} for action in actions
    )


def is_agent_feed_agents_md(path: Path) -> bool:
    if not path.is_file():
        return False
    text = path.read_text(encoding="utf-8")
    return (
        "AI Development Instructions" in text
        and "repository-level entry contract" in text
        and ".agents/rules/outcome-boundary.md" in text
    )


def is_agent_feed_agents_dir(path: Path) -> bool:
    readme = path / "README.md"
    return (
        path.is_dir()
        and readme.is_file()
        and "# AI Development Engineering" in readme.read_text(encoding="utf-8")
        and (path / "rules/outcome-boundary.md").is_file()
        and (path / "scripts/check-agent-assets.sh").is_file()
    )


def remove_empty_parents(path: Path) -> None:
    while path.name in {"rules", ".cursor", ".claude", ".codex"}:
        try:
            path.rmdir()
        except OSError:
            return
        path = path.parent


def _plan_file(
    actions: list[WriteAction],
    path: Path,
    *,
    action: str,
    safe: bool,
    unmanaged_detail: str,
    managed_detail: str,
) -> None:
    _plan_path(
        actions,
        path,
        action=action,
        safe=safe,
        unmanaged_detail=unmanaged_detail,
        managed_detail=managed_detail,
    )


def _plan_path(
    actions: list[WriteAction],
    path: Path,
    *,
    action: str,
    safe: bool,
    unmanaged_detail: str,
    managed_detail: str,
) -> None:
    if not path.exists():
        return
    if safe:
        actions.append(WriteAction(path, action, managed_detail))
    else:
        actions.append(WriteAction(path, "skip", unmanaged_detail))

File: src/node/src/uninstall.ts
import { existsSync, readFileSync, readdirSync, rmSync, rmdirSync, statSync, writeFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";

import { isManagedSkillMirror, isManagedCursorRule, missingClaudeSnippets } from "./adapters.js";
import { type WriteAction } from "./template.js";
import { checkAssetTrust } from "./trust.js";

export function uninstallPlan(root: string, dryRun: boolean): WriteAction[] {
  const action = dryRun ? "would delete" : "delete";
  const actions: WriteAction[] = [];

  planPath(actions, join(root, "CLAUDE.md"), {
    action,
    safe: isManagedClaudeMd(join(root, "CLAUDE.md")),
    unmanagedDetail: "unmanaged Claude instructions; not removed",
    managedDetail: "managed Claude adapter",
  });

  const claudeRoot = join(root, ".claude");
  const claudeSkills = join(claudeRoot, "skills");
  const claudeReadme = join(claudeRoot, "README.md");
  const managedClaudeSkills = isManagedSkillMirror(claudeRoot);
  planPath(actions, claudeSkills, {
    action,
    safe: managedClaudeSkills,
    unmanagedDetail: "unmanaged Claude skills; not removed",
    managedDetail: "managed Claude skills mirror",
  });
  planPath(actions, claudeReadme, {
    action,
    safe: managedClaudeSkills,
    unmanagedDetail: "unmanaged Claude README; not removed",
    managedDetail: "managed Claude skills README",
  });

  planPath(actions, join(root, ".cursor", "rules", "agent-feed.mdc"), {
    action,
    safe: isManagedCursorRule(join(root, ".cursor", "rules", "agent-feed.mdc")),
    unmanagedDetail: "unmanaged Cursor rule; not removed",
    managedDetail: "managed Cursor adapter",
  });

  const codexSkills = join(root, ".codex", "skills");
  const agentsSkills = join(root, ".agents", "skills");
  planPath(actions, codexSkills, {
    action,
    safe: existsSync(agentsSkills) && existsSync(codexSkills) && sameTree(agentsSkills, codexSkills),
    unmanagedDetail: "legacy .codex/skills is not a verified mirror; not removed",
    managedDetail: "legacy Agent Feed Codex skill mirror",
  });

  planPath(actions, join(root, "AGENTS.md"), {
    action,
    safe: isAgentFeedAgentsMd(join(root, "AGENTS.md")),
    unmanagedDetail: "unmanaged AGENTS.md; not removed",
    managedDetail: "Agent Feed entry contract",
  });

  planPath(actions, join(root, ".agents"), {
    action,
    safe: isAgentFeedAgentsDir(join(root, ".agents")),
    unmanagedDetail: "unmanaged .agents directory; not removed",
    managedDetail: "Agent Feed protocol directory",
  });

  actions.push(...projectTrustUninstallPlan(root, dryRun));
  return actions;
}

export function applyUninstallPlan(root: string, actions: WriteAction[]): WriteAction[] {
  const applied: WriteAction[] = [];
  const parentsToPrune = new Set<string>();
  for (const item of actions) {
    if (item.action === "update" && item.detail === "remove external trust state for this project") {
      removeProjectTrustState(root, item.path);
      applied.push({ path: item.path, action: "updated", detail: item.detail });
      continue;
    }
    if (item.action !== "delete") {
      continue;
    }
    if (!existsSync(item.path)) {
      applied.push({ path: item.path, action: "skip", detail: "already absent" });
      continue;
    }
    rmSync(item.path, { recursive: true, force: true });
    applied.push({ path: item.path, action: "deleted", detail: item.detail });
    if (["agent-feed.mdc", "README.md", "skills"].includes(basename(item.path))) {
      parentsToPrune.add(dirname(item.path));
    }
  }
  for (const directory of parentsToPrune) {
    removeEmptyParents(directory);
  }
  return applied;
}

export function uninstallHasDeletions(actions: WriteAction[]): boolean {
  return actions.some((item) =>
    item.action === "delete" ||
    item.action === "would delete" ||
    item.action === "update" ||
    item.action === "would update",
  );
}

function planPath(
  actions: WriteAction[],
  path: string,
  options: {
    action: string;
    safe: boolean;
    unmanagedDetail: string;
    managedDetail: string;
  },
): void {
  if (!existsSync(path)) {
    return;
  }
  actions.push({
    path,
    action: options.safe ? options.action : "skip",
    detail: options.safe ? options.managedDetail : options.unmanagedDetail,
  });
}

function isManagedClaudeMd(path: string): boolean {
  if (!existsSync(path)) {
    return false;
  }
  return missingClaudeSnippets(path).length === 0 && readFileSync(path, "utf8").includes("## Claude Code");
}

function isAgentFeedAgentsMd(path: string): boolean {
  if (!existsSync(path)) {
    return false;
  }
  const text = readFileSync(path, "utf8");
  return (
    text.includes("AI Development Instructions") &&
    text.includes("repository-level entry contract") &&
    text.includes(".agents/rules/outcome-boundary.md")
  );
}

function isAgentFeedAgentsDir(path: string): boolean {
  const readme = join(path, "README.md");
  return (
    existsSync(path) &&
    existsSync(readme) &&
    readFileSync(readme, "utf8").includes("# AI Development Engineering") &&
    existsSync(join(path, "rules", "outcome-boundary.md")) &&
    existsSync(join(path, "scripts", "check-agent-assets.sh"))
  );
}

function sameTree(left: string, right: string): boolean {
  if (!existsSync(left) || !existsSync(right)) {
    return false;
  }
  const leftFiles = walkTree(left).filter((item) => statSync(item).isFile());
  const rightFiles = walkTree(right).filter((item) => statSync(item).isFile());
  const leftRel = leftFiles.map((item) => item.slice(left.length + 1)).sort();
  const rightRel = rightFiles.map((item) => item.slice(right.length + 1)).sort();
  if (leftRel.join("\n") !== rightRel.join("\n")) {
    return false;
  }
  return leftRel.every(
    (relPath) => readFileSync(join(left, relPath), "utf8") === readFileSync(join(right, relPath), "utf8"),
  );
}

function walkTree(root: string): string[] {
  if (!existsSync(root)) {
    return [];
  }
  const results: string[] = [];
  for (const entry of readdirSync(root)) {
    if (entry === ".DS_Store" || entry === "__pycache__") {
      continue;
    }
    const next = join(root, entry);
    results.push(next);
    if (statSync(next).isDirectory()) {
      results.push(...walkTree(next));
    }
  }
  return results;
}

function projectTrustUninstallPlan(root: string, dryRun: boolean): WriteAction[] {
  const report = checkAssetTrust(root);
  if (!report.configPath || report.errors.length > 0 || report.missingState) {
    return [];
  }
  const configText = readFileSync(report.configPath, "utf8");
  const parsed = JSON.parse(configText) as { projects?: Record<string, unknown> };
  const projects = parsed.projects;
  const key = resolve(root);
  if (!projects || typeof projects !== "object" || Array.isArray(projects) || !(key in projects)) {
    return [];
  }
  return [
    {
      path: report.configPath,
      action: dryRun ? "would update" : "update",
      detail: "remove external trust state for this project",
    },
  ];
}

function removeProjectTrustState(root: string, configPath: string): void {
  const parsed = JSON.parse(readFileSync(configPath, "utf8")) as { projects?: Record<string, unknown> };
  if (parsed.projects && typeof parsed.projects === "object" && !Array.isArray(parsed.projects)) {
    delete parsed.projects[resolve(root)];
  }
  writeFileSync(configPath, `${JSON.stringify(parsed, null, 2)}\n`, "utf8");
}

function removeEmptyParents(path: string): void {
  let current = path;
  while (["rules", ".cursor", ".claude", ".codex"].includes(basename(current))) {
    try {
      rmdirSync(current);
    } catch {
      return;
    }
    current = dirname(current);
  }
}
